File: identity/src/main/java/androidx/security/identity/GattServer.java
/*
 * Copyright 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.security.identity;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattServer;
import android.bluetooth.BluetoothGattServerCallback;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothManager;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import androidx.security.identity.Constants.LoggingFlag;
import java.io.ByteArrayOutputStream;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.UUID;

class GattServer extends BluetoothGattServerCallback {
    private static final String TAG = "GattServer";
    private final byte[] mEncodedEDeviceKeyBytes;
    private final @LoggingFlag int mLoggingFlags;

    UUID mCharacteristicStateUuid;
    UUID mCharacteristicClient2ServerUuid;
    UUID mCharacteristicServer2ClientUuid;
    UUID mCharacteristicIdentUuid;

    // This is what the 16-bit UUID 0x29 0x02 is encoded like.
    UUID mClientCharacteristicConfigUuid =  UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");

    private final UUID mServiceUuid;
    private final BluetoothManager mBluetoothManager;
    private final Context mContext;
    private BluetoothGattServer mGattServer;
    Listener mListener;
    boolean mInhibitCallbacks = false;

    private BluetoothDevice mCurrentConnection;
    private int mNegotiatedMtu = 0;

    BluetoothGattCharacteristic mCharacteristicState;
    BluetoothGattCharacteristic mCharacteristicClient2Server;
    BluetoothGattCharacteristic mCharacteristicServer2Client;
    BluetoothGattCharacteristic mCharacteristicIdent;
    private byte[] mIdentValue;

    GattServer(@NonNull Context context, @LoggingFlag int loggingFlags, @NonNull BluetoothManager bluetoothManager,
        @NonNull UUID serviceUuid, @NonNull byte[] encodedEDeviceKeyBytes,
        @NonNull UUID characteristicStateUuid,
        @NonNull UUID characteristicClient2ServerUuid,
        @NonNull UUID characteristicServer2ClientUuid,
        @Nullable UUID characteristicIdentUuid) {
        mContext = context;
        Log.i(TAG, "LoggingFlags: " + loggingFlags);
        mLoggingFlags = loggingFlags;
        mBluetoothManager = bluetoothManager;
        mServiceUuid = serviceUuid;
        mEncodedEDeviceKeyBytes = encodedEDeviceKeyBytes;
        mCharacteristicStateUuid = characteristicStateUuid;
        mCharacteristicClient2ServerUuid = characteristicClient2ServerUuid;
        mCharacteristicServer2ClientUuid = characteristicServer2ClientUuid;
        mCharacteristicIdentUuid = characteristicIdentUuid;
    }

    void setListener(@Nullable Listener listener) {
        mListener = listener;
    }

    boolean start() {
        byte[] ikm = mEncodedEDeviceKeyBytes;
        byte[] info = new byte[] {'B', 'L', 'E', 'I', 'd', 'e', 'n', 't'};
        byte[] salt = new byte[] {};
        mIdentValue = Util.computeHkdf("HmacSha256", ikm, salt, info, 16);

        mGattServer = mBluetoothManager.openGattServer(mContext, this);
        if (mGattServer == null) {
            return false;
        }

        BluetoothGattService service = new BluetoothGattService(mServiceUuid,
                BluetoothGattService.SERVICE_TYPE_PRIMARY);

        BluetoothGattCharacteristic c;
        BluetoothGattDescriptor d;

        // State
        c = new BluetoothGattCharacteristic(mCharacteristicStateUuid,
                BluetoothGattCharacteristic.PROPERTY_NOTIFY
                        | BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
                BluetoothGattCharacteristic.PERMISSION_WRITE);
        d = new BluetoothGattDescriptor(mClientCharacteristicConfigUuid,
                BluetoothGattDescriptor.PERMISSION_WRITE);
        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE);
        c.addDescriptor(d);
        service.addCharacteristic(c);
        mCharacteristicState = c;

        // Client2Server
        c = new BluetoothGattCharacteristic(mCharacteristicClient2ServerUuid,
                BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
                BluetoothGattCharacteristic.PERMISSION_WRITE);
        service.addCharacteristic(c);
        mCharacteristicClient2Server = c;

        // Server2Client
        c = new BluetoothGattCharacteristic(mCharacteristicServer2ClientUuid,
                BluetoothGattCharacteristic.PROPERTY_NOTIFY,
                BluetoothGattCharacteristic.PERMISSION_WRITE);
        d = new BluetoothGattDescriptor(mClientCharacteristicConfigUuid,
                BluetoothGattDescriptor.PERMISSION_WRITE);
        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE);
        c.addDescriptor(d);
        service.addCharacteristic(c);
        mCharacteristicServer2Client = c;

        // Ident
        if (mCharacteristicIdentUuid != null) {
            c = new BluetoothGattCharacteristic(mCharacteristicIdentUuid,
                BluetoothGattCharacteristic.PROPERTY_READ,
                BluetoothGattCharacteristic.PERMISSION_READ);
            service.addCharacteristic(c);
            mCharacteristicIdent = c;
        }

        mGattServer.addService(service);

        return true;
    }

    void stop() {
        mInhibitCallbacks = true;
        if (mGattServer != null) {
            if (mCurrentConnection != null) {
                mGattServer.cancelConnection(mCurrentConnection);
            }
            mGattServer.close();
        }
    }

    @Override
    public void onConnectionStateChange(BluetoothDevice device, int status, int newState) {
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
            Log.i(TAG, "onConnectionStateChange: " + status + " " + newState);
        }
        if (newState == BluetoothProfile.STATE_CONNECTED) {
            if (mCurrentConnection != null) {
                Log.w(TAG, "Got a connection but we already have one");
                // TODO: possible to disconnect that 2nd device?
                return;
            }
            mCurrentConnection = device;

        } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
            if (!device.getAddress().equals(mCurrentConnection.getAddress())) {
                Log.w(TAG, "Got a disconnection from a device we haven't seen before");
                return;
            }
            mCurrentConnection = null;
            reportPeerDisconnected();
        }
    }

    @Override
    public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, int offset,
            BluetoothGattCharacteristic characteristic) {
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
            Log.i(TAG, "onCharacteristicReadRequest: " + requestId + " " + offset + " " + characteristic.getUuid());
        }
        if (mCharacteristicIdentUuid != null
            && characteristic.getUuid().equals(mCharacteristicIdentUuid)) {
            mGattServer.sendResponse(device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    0,
                    mIdentValue);
        } else {
            reportError(new Error("Read on unexpected characteristic with UUID "
                    + characteristic.getUuid()));
        }
    }

    ByteArrayOutputStream mIncomingMessage = new ByteArrayOutputStream();

    @Override
    public void onCharacteristicWriteRequest(BluetoothDevice device,
            int requestId,
            BluetoothGattCharacteristic characteristic,
            boolean preparedWrite,
            boolean responseNeeded,
            int offset,
            byte[] value) {
        UUID charUuid = characteristic.getUuid();
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC_VERBOSE) != 0) {
            Log.i(TAG, "onCharacteristicWriteRequest: "
                + characteristic.getUuid() + " " + offset + " " + Util.toHex(value));
        }
        if (charUuid.equals(mCharacteristicStateUuid)
            && value.length == 1 && value[0] == 0x01) {
            reportPeerConnected();
        } else if (charUuid.equals(mCharacteristicClient2ServerUuid)) {
            if (value.length < 1) {
                reportError(new Error("Invalid value with length " + value.length));
                return;
            }
            mIncomingMessage.write(value, 1, value.length - 1);
            if (value[0] == 0x00) {
                // Last message.
                byte[] entireMessage = mIncomingMessage.toByteArray();
                mIncomingMessage.reset();
                reportMessageReceived(entireMessage);
            }

            if (responseNeeded) {
                mGattServer.sendResponse(device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        0,
                        null);
            }
        } else {
            reportError(new Error("Write on unexpected characteristic with UUID "
                    + characteristic.getUuid()));
        }
    }

    @Override
    public void onDescriptorReadRequest(BluetoothDevice device, int requestId, int offset,
            BluetoothGattDescriptor descriptor) {
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC_VERBOSE) != 0) {
            Log.i(TAG, "onDescriptorReadRequest: "
                + descriptor.getCharacteristic().getUuid() + " " + offset);
        }
        /* Do nothing */
    }

    @Override
    public void onDescriptorWriteRequest(BluetoothDevice device, int requestId,
            BluetoothGattDescriptor descriptor,
            boolean preparedWrite, boolean responseNeeded,
            int offset, byte[] value) {
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC_VERBOSE) != 0) {
            Log.i(TAG, "onDescriptorWriteRequest: "
                + descriptor.getCharacteristic().getUuid() + " " + offset + " " + Util.toHex(value));
        }
        if (responseNeeded) {
            mGattServer.sendResponse(device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    0,
                    null);
        }
    }

    @Override
    public void onMtuChanged(BluetoothDevice device, int mtu) {
        mNegotiatedMtu = mtu;
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
            Log.i(TAG, "Negotiated MTU " + mtu);
        }
    }

    Queue<byte[]> mWritingQueue = new ArrayDeque<>();
    boolean writeIsOutstanding = false;

    void drainWritingQueue() {
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
            Log.i(TAG, "drainWritingQueue " + writeIsOutstanding);
        }
        if (writeIsOutstanding) {
            return;
        }
        byte[] chunk = mWritingQueue.poll();
        if (chunk == null) {
            return;
        }

        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC_VERBOSE) != 0) {
            Log.i(TAG, "writing chunk to " + mCharacteristicServer2Client.getUuid() + " " + Util.toHex(chunk));
        }

        mCharacteristicServer2Client.setValue(chunk);
        if (!mGattServer.notifyCharacteristicChanged(mCurrentConnection,
                mCharacteristicServer2Client, false)) {
            reportError(new Error("Error calling notifyCharacteristicsChanged on Server2Client"));
            return;
        }
        writeIsOutstanding = true;
    }

    @Override
    public void onNotificationSent(BluetoothDevice device, int status) {
        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
            Log.i(TAG, "onNotificationSent " + status);
        }
        if (status != BluetoothGatt.GATT_SUCCESS) {
            reportError(new Error("Error in onNotificationSent status=" + status));
            return;
        }
        writeIsOutstanding = false;
        drainWritingQueue();
    }

    void sendMessage(@NonNull byte[] data) {
        if (mNegotiatedMtu == 0) {
            Log.w(TAG, "MTU not negotiated, defaulting to 23. Performance will suffer.");
            mNegotiatedMtu = 23;
        }

        if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC_VERBOSE) != 0) {
            Log.i(TAG, "sendMessage " + Util.toHex(data));
        }

        // Three less the MTU but we also need room for the leading 0x00 or 0x01.
        //
        int maxChunkSize = mNegotiatedMtu - 4;

        int offset = 0;
        do {
            boolean moreChunksComing = (offset + maxChunkSize < data.length);
            int size = data.length - offset;
            if (size > maxChunkSize) {
                size = maxChunkSize;
            }

            byte[] chunk = new byte[size + 1];
            chunk[0] = moreChunksComing ? (byte) 0x01 : (byte) 0x00;
            System.arraycopy(data, offset, chunk, 1, size);

            mWritingQueue.add(chunk);

            offset += size;
        } while (offset < data.length);
        drainWritingQueue();
    }

    void reportPeerConnected() {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onPeerConnected();
        }
    }

    void reportPeerDisconnected() {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onPeerDisconnected();
        }
    }

    void reportMessageReceived(@NonNull byte[] data) {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onMessageReceived(data);
        }
    }

    void reportError(@NonNull Throwable error) {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onError(error);
        }
    }

    interface Listener {
        void onPeerConnected();
        void onPeerDisconnected();
        void onMessageReceived(@NonNull byte[] data);
        void onError(@NonNull Throwable error);
    }

    public void sendTransportSpecificTermination() {
        byte[] terminationCode = new byte[]{(byte) 0x02};
        mCharacteristicState.setValue(terminationCode);
        if (!mGattServer.notifyCharacteristicChanged(mCurrentConnection,
                mCharacteristicState, false)) {
            reportError(new Error("Error calling notifyCharacteristicsChanged on State"));
        }
    }
}

File: identity-android/src/main/java/com/android/identity/android/mdoc/transport/GattServer.kt
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.android.mdoc.transport

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattServer
import android.bluetooth.BluetoothGattServerCallback
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.os.Build
import com.android.identity.crypto.Algorithm
import com.android.identity.crypto.Crypto
import com.android.identity.util.Logger
import com.android.identity.util.toHex
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.ArrayDeque
import java.util.Queue
import java.util.UUID

@Suppress("deprecation")
@SuppressLint("MissingPermission")
internal class GattServer(
    private val context: Context,
    private val bluetoothManager: BluetoothManager,
    private val serviceUuid: UUID,
    private val encodedEDeviceKeyBytes: ByteArray?,
    var characteristicStateUuid: UUID,
    var characteristicClient2ServerUuid: UUID,
    var characteristicServer2ClientUuid: UUID,
    var characteristicIdentUuid: UUID?,
    var characteristicL2CAPUuid: UUID?
) : BluetoothGattServerCallback() {
    var psm: Int? = null
        private set
    var listener: Listener? = null

    // This is what the 16-bit UUID 0x29 0x02 is encoded like.
    private var clientCharacteristicConfigUuid = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    private var inhibitCallbacks = false
    private var characteristicState: BluetoothGattCharacteristic? = null
    private var characteristicClient2Server: BluetoothGattCharacteristic? = null
    private var characteristicServer2Client: BluetoothGattCharacteristic? = null
    private var characteristicIdent: BluetoothGattCharacteristic? = null
    private var characteristicL2CAP: BluetoothGattCharacteristic? = null
    private var incomingMessage = ByteArrayOutputStream()
    private var writingQueue: Queue<ByteArray> = ArrayDeque()
    private var gattServer: BluetoothGattServer? = null
    private var currentConnection: BluetoothDevice? = null
    private var negotiatedMtu = 0
    private var l2capServer: L2CAPServer? = null
    private var identValue: ByteArray? = null
    private var usingL2CAP = false

    fun setEDeviceKeyBytes(encodedEDeviceKeyBytes: ByteArray) {
        val ikm: ByteArray = encodedEDeviceKeyBytes
        val info = "BLEIdent".toByteArray()
        val salt = byteArrayOf()
        identValue = Crypto.hkdf(Algorithm.HMAC_SHA256, ikm, salt, info, 16)
    }

    @SuppressLint("NewApi")
    fun start(): Boolean {
        if (encodedEDeviceKeyBytes != null) {
            setEDeviceKeyBytes(encodedEDeviceKeyBytes)
        }
        gattServer = try {
            bluetoothManager.openGattServer(context, this)
        } catch (e: SecurityException) {
            reportError(e)
            return false
        }
        if (gattServer == null) {
            return false
        }
        val service = BluetoothGattService(
            serviceUuid,
            BluetoothGattService.SERVICE_TYPE_PRIMARY
        )
        var c: BluetoothGattCharacteristic
        var d: BluetoothGattDescriptor

        // State
        c = BluetoothGattCharacteristic(
            characteristicStateUuid, BluetoothGattCharacteristic.PROPERTY_NOTIFY
                    or BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        )
        d = BluetoothGattDescriptor(
            clientCharacteristicConfigUuid,
            BluetoothGattDescriptor.PERMISSION_WRITE
        )
        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE)
        c.addDescriptor(d)
        service.addCharacteristic(c)
        characteristicState = c

        // Client2Server
        c = BluetoothGattCharacteristic(
            characteristicClient2ServerUuid,
            BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        )
        service.addCharacteristic(c)
        characteristicClient2Server = c

        // Server2Client
        c = BluetoothGattCharacteristic(
            characteristicServer2ClientUuid,
            BluetoothGattCharacteristic.PROPERTY_NOTIFY,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        )
        d = BluetoothGattDescriptor(
            clientCharacteristicConfigUuid,
            BluetoothGattDescriptor.PERMISSION_WRITE
        )
        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE)
        c.addDescriptor(d)
        service.addCharacteristic(c)
        characteristicServer2Client = c

        // Ident
        if (characteristicIdentUuid != null) {
            c = BluetoothGattCharacteristic(
                characteristicIdentUuid,
                BluetoothGattCharacteristic.PROPERTY_READ,
                BluetoothGattCharacteristic.PERMISSION_READ
            )
            service.addCharacteristic(c)
            characteristicIdent = c
        }

        // Offers support to L2CAP when we have UUID characteristic and the OS version support it
        usingL2CAP =
            characteristicL2CAPUuid != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
        Logger.i(TAG, "Is L2CAP supported: $usingL2CAP")
        if (usingL2CAP) {
            // Start L2CAP socket server
            l2capServer = L2CAPServer(object : L2CAPServer.Listener {
                override fun onPeerConnected() {
                    reportPeerConnected()
                }

                override fun onPeerDisconnected() {
                    reportPeerDisconnected()
                }

                override fun onMessageReceived(data: ByteArray) {
                    reportMessageReceived(data)
                }

                override fun onError(error: Throwable) {
                    reportError(error)
                }
            })
            psm = l2capServer!!.start(bluetoothManager.adapter)
            if (psm == null) {
                Logger.w(TAG, "Error starting L2CAP server")
                l2capServer = null
                usingL2CAP = false
            } else {
                Logger.i(TAG, "Listening on L2CAP with PSM ${psm}")
                c = BluetoothGattCharacteristic(
                    characteristicL2CAPUuid,
                    BluetoothGattCharacteristic.PROPERTY_READ,
                    BluetoothGattCharacteristic.PERMISSION_READ
                )
                service.addCharacteristic(c)
                characteristicL2CAP = c
            }
        }
        try {
            gattServer!!.addService(service)
        } catch (e: SecurityException) {
            reportError(e)
            return false
        }
        return true
    }

    fun stop() {
        Logger.i(TAG, "stop")
        inhibitCallbacks = true
        if (l2capServer != null) {
            l2capServer!!.stop()
            l2capServer = null
        }
        if (gattServer != null) {
            // used to convey we want to shutdown once all write are done.
            sendMessage(ByteArray(0))
        }
    }

    override fun onConnectionStateChange(device: BluetoothDevice, status: Int, newState: Int) {
        Logger.d(TAG, "onConnectionStateChange: ${device.address} $status + $newState")
        if (newState == BluetoothProfile.STATE_DISCONNECTED && currentConnection != null &&
            device.address == currentConnection!!.address) {
            Logger.d(TAG, "Device ${currentConnection!!.address} which we're currently " +
                        "connected to, has disconnected")
            currentConnection = null
            reportPeerDisconnected()
        }
    }

    @SuppressLint("NewApi")
    override fun onCharacteristicReadRequest(
        device: BluetoothDevice, requestId: Int, offset: Int,
        characteristic: BluetoothGattCharacteristic
    ) {
        Logger.d(TAG, "onCharacteristicReadRequest: ${device.address} $requestId " +
                "$offset ${characteristic.uuid}")
        if (characteristicIdentUuid != null && characteristic.uuid == characteristicIdentUuid) {
            try {
                val ident = if (identValue != null) identValue!! else ByteArray(0)
                gattServer!!.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    0,
                    ident
                )
            } catch (e: SecurityException) {
                reportError(e)
            }
        } else if (characteristicL2CAP != null && characteristic.uuid == characteristicL2CAPUuid) {
            if (!usingL2CAP) {
                reportError(Error("Unexpected read request for L2CAP characteristic, not supported"))
                return
            }
            // TODO: it's not clear this is the right way to encode the PSM and 18013-5 doesn't
            //   seem to give enough guidance on it.
            val encodedPsmValue = ByteBuffer.allocate(4).putInt(psm!!).array()
            gattServer!!.sendResponse(
                device,
                requestId,
                BluetoothGatt.GATT_SUCCESS,
                0,
                encodedPsmValue
            )
        } else {
            reportError(
                Error(
                    "Read on unexpected characteristic with UUID "
                            + characteristic.uuid
                )
            )
        }
    }

    override fun onCharacteristicWriteRequest(
        device: BluetoothDevice,
        requestId: Int,
        characteristic: BluetoothGattCharacteristic,
        preparedWrite: Boolean,
        responseNeeded: Boolean,
        offset: Int,
        value: ByteArray
    ) {
        val charUuid = characteristic.uuid
        Logger.d(TAG, "onCharacteristicWriteRequest: ${device.address} $requestId " +
                "$offset ${characteristic.uuid} ${value.toHex()}")

        // If we are connected to a device, ignore write from any other device
        if (currentConnection != null &&
            device.address != currentConnection!!.address
        ) {
            Logger.e(TAG, "Ignoring characteristic write request from ${device.address} " +
                    "since we're already connected to ${currentConnection!!.address}")
            return
        }
        if (charUuid == characteristicStateUuid && value.size == 1) {
            if (value[0].toInt() == 0x01) {
                // Close server socket when the connection was done by state characteristic
                stopL2CAPServer()
                if (currentConnection != null) {
                    Logger.e(TAG, "Ignoring connection attempt from ${device.address} " +
                                "since we're already connected to ${currentConnection!!.address}")
                } else {
                    currentConnection = device
                    Logger.d(TAG, "Received connection (state 0x01 on State characteristic) "
                                + "from ${currentConnection!!.address}")
                }
                reportPeerConnected()
            } else if (value[0].toInt() == 0x02) {
                reportTransportSpecificSessionTermination()
            } else {
                reportError(Error("Invalid byte ${value[0]} for state characteristic"))
            }
        } else if (charUuid == characteristicClient2ServerUuid) {
            if (value.size < 1) {
                reportError(Error("Invalid value with length ${value.size}"))
                return
            }
            if (currentConnection == null) {
                // We expect a write 0x01 on the State characteristic before we consider
                // the device to be connected.
                reportError(Error("Write on Client2Server but not connected yet"))
                return
            }
            incomingMessage.write(value, 1, value.size - 1)
            val isLast = (value[0].toInt() == 0x00)
            Logger.dHex(TAG, "Received chunk with ${value.size} bytes " +
                    "(last=$isLast), incomingMessage.length=${incomingMessage.toByteArray().size}", value)
            if (value[0].toInt() == 0x00) {
                // Last message.
                val entireMessage = incomingMessage.toByteArray()
                incomingMessage.reset()
                reportMessageReceived(entireMessage)
            } else if (value[0].toInt() == 0x01) {
                if (value.size != characteristicValueSize) {
                    Logger.w(TAG,
                        "Client2Server received ${value.size} bytes which is not the " +
                            "expected $characteristicValueSize bytes"
                    )
                    return
                }
            } else {
                reportError(Error(
                    "Invalid first byte ${value[0]} in Client2Server data chunk, expected 0 or 1"))
                return
            }
            if (responseNeeded) {
                try {
                    gattServer!!.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        0,
                        null
                    )
                } catch (e: SecurityException) {
                    reportError(e)
                }
            }
        } else {
            reportError(Error(
                "Write on unexpected characteristic with UUID ${characteristic.uuid}")
            )
        }
    }

    private fun stopL2CAPServer() {
        // Close server socket when the connection was done by state characteristic
        if (l2capServer != null) {
            l2capServer!!.stop()
            l2capServer = null
        }
        usingL2CAP = false
    }

    override fun onDescriptorReadRequest(
        device: BluetoothDevice, requestId: Int, offset: Int,
        descriptor: BluetoothGattDescriptor
    ) {
        Logger.d(TAG, "onDescriptorReadRequest: ${device.address} " +
                "${descriptor.characteristic.uuid} ${descriptor.characteristic.uuid} $offset")
        /* Do nothing */
    }

    override fun onDescriptorWriteRequest(
        device: BluetoothDevice, requestId: Int,
        descriptor: BluetoothGattDescriptor,
        preparedWrite: Boolean, responseNeeded: Boolean,
        offset: Int, value: ByteArray
    ) {
        if (Logger.isDebugEnabled) {
            Logger.d(
                TAG, "onDescriptorWriteRequest: ${device.address}" +
                        "${descriptor.characteristic.uuid} $offset ${value.toHex()}"
            )
        }
        if (responseNeeded) {
            try {
                gattServer!!.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    0,
                    null
                )
            } catch (e: SecurityException) {
                reportError(e)
            }
        }
    }

    override fun onMtuChanged(device: BluetoothDevice, mtu: Int) {
        negotiatedMtu = mtu
        Logger.d(TAG, "Negotiated MTU $mtu for $${device.address}")
    }

    private var characteristicValueSizeMemoized = 0
    private val characteristicValueSize: Int
        get() {
            if (characteristicValueSizeMemoized > 0) {
                return characteristicValueSizeMemoized
            }
            var mtuSize = negotiatedMtu
            if (mtuSize == 0) {
                Logger.w(TAG, "MTU not negotiated, defaulting to 23. Performance will suffer.")
                mtuSize = 23
            }
            characteristicValueSizeMemoized = DataTransportBle.bleCalculateAttributeValueSize(mtuSize)
            return characteristicValueSizeMemoized
        }

    fun drainWritingQueue() {
        Logger.d(TAG, "drainWritingQueue")
        val chunk = writingQueue.poll() ?: return
        if (chunk.size == 0) {
            Logger.d(TAG, "Chunk is length 0, shutting down GattServer in 1000ms")
            // TODO: On some devices we lose messages already sent if we don't have a delay like
            //  this. Need to properly investigate if this is a problem in our stack or the
            //  underlying BLE subsystem.
            Thread.sleep(1000)
            Logger.d(TAG, "Shutting down GattServer now")
            try {
                if (currentConnection != null) {
                    gattServer!!.cancelConnection(currentConnection)
                }
                gattServer!!.close()
            } catch (e: SecurityException) {
                Logger.e(TAG, "Caught SecurityException while shutting down", e)
            }
            gattServer = null
            return
        }
        val isLast = chunk[0].toInt() == 0x00
        Logger.d(TAG, "Sending chunk with ${chunk.size} bytes (last=$isLast)")
        characteristicServer2Client!!.setValue(chunk)
        try {
            if (!gattServer!!.notifyCharacteristicChanged(
                    currentConnection,
                    characteristicServer2Client, false
                )
            ) {
                reportError(
                    Error("Error calling notifyCharacteristicsChanged on Server2Client")
                )
                return
            }
        } catch (e: SecurityException) {
            reportError(e)
            return
        }
    }

    override fun onNotificationSent(device: BluetoothDevice, status: Int) {
        Logger.d(TAG, "onNotificationSent $status for ${device.address}")
        if (status != BluetoothGatt.GATT_SUCCESS) {
            reportError(Error("Error in onNotificationSent status=$status"))
            return
        }
        drainWritingQueue()
    }

    fun sendMessage(data: ByteArray) {
        Logger.dHex(TAG, "sendMessage", data)

        // Uses socket L2CAP when it is available
        if (l2capServer != null) {
            l2capServer!!.sendMessage(data)
            return
        }
        // Only initiate the write if no other write was outstanding.
        val queueNeedsDraining = (writingQueue.size == 0)
        if (data.size == 0) {
            // Data of length 0 is used to signal we should shut down.
            writingQueue.add(data)
        } else {
            // Also need room for the leading 0x00 or 0x01.
            //
            val maxDataSize = characteristicValueSize - 1
            var offset = 0
            do {
                val moreDataComing = offset + maxDataSize < data.size
                var size = data.size - offset
                if (size > maxDataSize) {
                    size = maxDataSize
                }
                val chunk = ByteArray(size + 1)
                chunk[0] = if (moreDataComing) 0x01.toByte() else 0x00.toByte()
                System.arraycopy(data, offset, chunk, 1, size)
                writingQueue.add(chunk)
                offset += size
            } while (offset < data.size)
        }
        if (queueNeedsDraining) {
            drainWritingQueue()
        }
    }

    fun reportPeerConnected() {
        if (listener != null && !inhibitCallbacks) {
            listener!!.onPeerConnected()
        }
    }

    fun reportPeerDisconnected() {
        if (listener != null && !inhibitCallbacks) {
            listener!!.onPeerDisconnected()
        }
    }

    fun reportMessageReceived(data: ByteArray) {
        if (listener != null && !inhibitCallbacks) {
            listener!!.onMessageReceived(data)
        }
    }

    fun reportError(error: Throwable) {
        if (listener != null && !inhibitCallbacks) {
            listener!!.onError(error)
        }
    }

    fun reportTransportSpecificSessionTermination() {
        if (listener != null && !inhibitCallbacks) {
            listener!!.onTransportSpecificSessionTermination()
        }
    }

    // When using L2CAP it doesn't support characteristics notification
    fun supportsTransportSpecificTerminationMessage(): Boolean {
        return !usingL2CAP
    }

    fun sendTransportSpecificTermination() {
        val terminationCode = byteArrayOf(0x02.toByte())
        characteristicState!!.setValue(terminationCode)
        try {
            if (!gattServer!!.notifyCharacteristicChanged(
                    currentConnection,
                    characteristicState, false
                )
            ) {
                reportError(Error("Error calling notifyCharacteristicsChanged on State"))
            }
        } catch (e: SecurityException) {
            reportError(e)
        }
    }

    internal interface Listener {
        fun onPeerConnected()
        fun onPeerDisconnected()
        fun onMessageReceived(data: ByteArray)
        fun onTransportSpecificSessionTermination()
        fun onError(error: Throwable)
    }

    companion object {
        private const val TAG = "GattServer"
    }
}
